File: OOP/Assignments & Projects/Practice Projects/tic-tac-toe.cpp
#include <iostream>
#include <limits> // for numeric_limits

#define ROWS 3
#define COLS 3

using namespace std;

char BOARD[ROWS][COLS]{};

int getInt(const string& prompt); // a good example for taking in input the right way
void print_board();
bool legality(int row, int col);
bool win_check(char symbol);

int main(){
    int input;
    cout << "Welcome to tic-tac-toe!!!" << endl;
    cout << "Menu" << endl;
    cout << "Press 1 to Play" << endl;
    cout << "Press 2 to Exit (any input)" << endl;

    bool flag = true;
    while(flag){
        switch(input){
            case 2:
                cout << "Bye!" << endl;
                return 0;
            case 1:
                cout << "Game on!" << endl;
                break;
            default:
                cout << "Not valid" << endl;
                flag = false;
            }
    }
    
    bool turn = false;
    char symbols[2] = {'X', 'O'};
    while(true){
        cout << "---------------------------" << endl;
        cout << symbols[turn] << "'s turn!" << endl;
        cout << "---------------------------" << endl;
        print_board();
        cout << "---------------------------" << endl;
        int row = getInt("ROW: ") - 1; //cout << "ROW: "; cin >> row;
        int col = getInt("COL: ") - 1; //cout << "COL: "; cin >> col;

        if (legality(row, col)){
            BOARD[row][col] = symbols[turn];
            if(win_check(symbols[turn])){
                print_board();
                cout << symbols[turn] << "'s wins!!!" << endl;
                break;
            }
            turn = not(turn);
        }
    }
}

// Special input function that controls for
int getInt(const string& prompt) {
    int value;
    while (true) {
        cout << prompt;
        cin >> value;

        if (cin.fail()) { // this line checks if you have mistakenly taken wrong input (maybe a string)
            cin.clear(); // clear error flags
            cin.ignore(numeric_limits<streamsize>::max(), '\n'); // discard invalid input
            cout << "Invalid input. Please enter an integer." << endl;
        } else {
            cin.ignore(numeric_limits<streamsize>::max(), '\n'); // clear extra characters
            return value;
        }
    }
}

void print_board(){
    for(int r = 0; r < ROWS; r++){
        for(int c = 0; c < COLS; c++){
            cout << BOARD[r][c] << " ";
        }
        cout << endl;
    }
}

bool legality(int row, int col){
    if(row > ROWS || col > ROWS){
        cout << "Outside of range!" << endl;
        return false;
    }
    if(row < 0 || col < 0){
        cout << "Outside of range!" << endl;
        return false;
    }
    if(BOARD[row][col] != '.'){
        cout << "Already taken!" << endl;
        return false;
    }
    return true;
}

bool win_check(char symbol){
    // Check rows and columns
    for (int i = 0; i < ROWS; i++) {
        if (BOARD[i][0] == symbol && BOARD[i][1] == symbol && BOARD[i][2] == symbol) return true;
        if (BOARD[0][i] == symbol && BOARD[1][i] == symbol && BOARD[2][i] == symbol) return true;
    }
    // Check diagonals
    if (BOARD[0][0] == symbol && BOARD[1][1] == symbol && BOARD[2][2] == symbol) return true;
    if (BOARD[0][2] == symbol && BOARD[1][1] == symbol && BOARD[2][0] == symbol) return true;
    return false;
}

File: OOP/Assignments & Projects/Practice Projects/tic-tac-toe.py
print('Tic Tac Toe!')
board = [
    ['.','.','.'],
    ['.','.','.'],
    ['.','.','.']
    ]

turn = False
symbols = ['X','O'] # X starts first

def print_board():
    for r in range(3):
        for c in range(3):
            print(board[r][c], end = ' ')
        print('')

def legality(row, col):
    if (row > 2) or (col > 2):
        print('ILLEGAL-1')
        return False
    if (row < 0) or (col < 0):
        print('ILLEGAL-2')
        return False
    if board[row][col] != '.':
        print('ILLEGAL-3')
        return False
    
    return True

def win_check():
    if board[0][0] == '.':
        return False
    else:
        check_1 = board[0][1] == board[0][0]
        check_2 = board[0][2] == board[0][0]
        if check_1 and check_2: return True

        check_3 = board[1][1] == board[0][0]
        check_4 = board[2][2] == board[0][0]
        if check_3 and check_4: return True

        check_5 = board[1][0] == board[0][0]
        check_6 = board[2][0] == board[0][0]
        if check_5 and check_6: return True

    if board[0][1] == '.':
        return False
    else:
        check_1 = board[1][1] == board[0][1]
        check_2 = board[2][1] == board[0][1]
        if check_1 and check_2: return True
    
    if board[0][2] == '.':
        return False
    else:
        check_1 = board[0][1] == board[0][2]
        check_2 = board[0][0] == board[0][2]
        if check_1 and check_2: return True

        check_3 = board[1][1] == board[0][2]
        check_4 = board[2][0] == board[0][2]
        if check_3 and check_4: return True

        check_5 = board[1][2] == board[0][2]
        check_6 = board[2][2] == board[0][2]
        if check_5 and check_6: return True
    
    if board[1][0] == '.':
        return False
    else:
        check_1 = board[1][1] == board[1][0]
        check_2 = board[1][2] == board[1][0]
        if check_1 and check_2: return True
    
    if board[2][0] == '.':
        return False
    else:
        check_1 = board[2][1] == board[2][0]
        check_2 = board[2][2] == board[2][0]
        if check_1 and check_2: return True
    
    return False

while (True):
    print("---------------------------")
    print(f'{symbols[turn]}\'s turn!')
    print("---------------------------")
    print_board()
    print("---------------------------")
    row = int(input('ROW: ')) # input by default is str
    col = int(input('COL: '))
    # the above line can give error if user inputs a string like "yes"
    # can be fixed but not interested right now

    if legality(row, col):
        board[row][col] = symbols[turn]
        if win_check():
            print_board()
            print(f'{symbols[turn]}\'s wins!!!')
            break
        turn = not(turn)
    
